File: src/main/java/mysql/hibernate/HibernateManager.java
package mysql.hibernate;

import core.MainLogger;
import mysql.hibernate.entity.GuildEntity;

import javax.persistence.EntityManagerFactory;
import javax.persistence.Persistence;
import java.util.HashMap;
import java.util.Map;

public class HibernateManager {

    private static EntityManagerFactory entityManagerFactory;

    public static void connect() {
        Map<String, Object> configOverrides = new HashMap<>();
        addConfigOverride(configOverrides, "hibernate.ogm.mongodb.host", "MONGODB_HOST");
        addConfigOverride(configOverrides, "hibernate.ogm.mongodb.port", "MONGODB_PORT");
        addConfigOverride(configOverrides, "hibernate.ogm.mongodb.username", "MONGODB_USER");
        addConfigOverride(configOverrides, "hibernate.ogm.mongodb.username", "MONGODB_PASSWORD");
        MainLogger.get().info("Connecting with MongoDB database");
        entityManagerFactory = Persistence.createEntityManagerFactory("lawliet", configOverrides);
    }

    public static EntityManagerWrapper createEntityManager() {
        return new EntityManagerWrapper(entityManagerFactory.createEntityManager());
    }

    public static GuildEntity findGuildEntity(long guildId) {
        return createEntityManager().findGuildEntity(guildId);
    }

    private static void addConfigOverride(Map<String, Object> configOverrides, String configKey, String environmentKey) {
        String value = System.getenv(environmentKey);
        if (value != null) {
            configOverrides.put(configKey, value);
        }
    }

}

File: src/main/java/mysql/hibernate/HibernateManager.kt
package mysql.hibernate

import core.MainLogger
import mysql.hibernate.entity.guild.GuildEntity
import mysql.hibernate.entity.user.UserEntity
import java.io.IOException
import java.util.*
import javax.persistence.EntityManagerFactory
import javax.persistence.Persistence

object HibernateManager {

    private lateinit var entityManagerFactory: EntityManagerFactory

    @JvmStatic
    @Throws(IOException::class)
    fun connect() {
        val props = Properties()
        props["hibernate.ogm.datastore.host"] = System.getenv("MONGODB_HOST")
        props["hibernate.ogm.datastore.username"] = System.getenv("MONGODB_USER")
        props["hibernate.ogm.datastore.password"] = System.getenv("MONGODB_PASSWORD")
        MainLogger.get().info("Connecting with MongoDB database")
        entityManagerFactory = Persistence.createEntityManagerFactory("lawliet", props)
    }

    @JvmStatic
    fun createEntityManager(callingClass: Class<*>): EntityManagerWrapper {
        return EntityManagerWrapper(entityManagerFactory.createEntityManager(), callingClass)
    }

    @JvmStatic
    fun findGuildEntity(guildId: Long, callingClass: Class<*>): GuildEntity {
        return createEntityManager(callingClass).findGuildEntity(guildId)
    }

    @JvmStatic
    fun findUserEntity(userId: Long, callingClass: Class<*>): UserEntity {
        return createEntityManager(callingClass).findUserEntity(userId)
    }

    @JvmStatic
    fun findUserEntityReadOnly(userId: Long, callingClass: Class<*>): UserEntity {
        return createEntityManager(callingClass).findUserEntityReadOnly(userId)
    }

}
